ImageBufferBackend has functions marked as const when they are not that in practice

kkinnunen-apple opened the pull request:
#### fc59b09d83136cf3bcec1fd91605a9ce8cd238ea
<pre>
ImageBufferBackend has functions marked as const when they are not that in practice
<a href="https://bugs.webkit.org/show_bug.cgi?id=253064">https://bugs.webkit.org/show_bug.cgi?id=253064</a>
rdar://106020570

Reviewed by Said Abou-Hallawa.

Remove const qualifier from ImageBufferBackend methods that are not
const in practice. Even getters like context() and get*() have implicit
assumptions about lazy work and caching, so it is not useful to pretend
the methods are const.

Remove ImageBufferBackend::platformLayer() const function altogether,
it is unused.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::platformLayer const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferBackend::getPixelBuffer):
(WebCore::ImageBufferBackend::copyNativeImageForDrawing const): Deleted.
(WebCore::ImageBufferBackend::getPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::copyToPlatformTexture):
(WebCore::ImageBufferBackend::platformLayer const): Deleted.
(WebCore::ImageBufferBackend::copyToPlatformTexture const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::context):
(WebCore::ImageBufferCGBitmapBackend::backendSize const):
(WebCore::ImageBufferCGBitmapBackend::copyNativeImage):
(WebCore::ImageBufferCGBitmapBackend::getPixelBuffer):
(WebCore::ImageBufferCGBitmapBackend::context const): Deleted.
(WebCore::ImageBufferCGBitmapBackend::copyNativeImage const): Deleted.
(WebCore::ImageBufferCGBitmapBackend::getPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::context):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::context const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableBitmapBackend::getPixelBuffer):
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage const): Deleted.
(WebKit::ImageBufferShareableBitmapBackend::getPixelBuffer const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::context):
(WebKit::ImageBufferRemoteIOSurfaceBackend::copyNativeImage):
(WebKit::ImageBufferRemoteIOSurfaceBackend::getPixelBuffer):
(WebKit::ImageBufferRemoteIOSurfaceBackend::context const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::copyNativeImage const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::getPixelBuffer const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::copyNativeImage):
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::copyNativeImage const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.h:

Canonical link: <a href="https://commits.webkit.org/260988@main">https://commits.webkit.org/260988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66fa8315f6efcaa19e44ba7b7507a44941a295b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42852 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1614 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119186 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114144 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20750 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10480 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102449 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115938 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43666 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30318 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85506 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11981 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31655 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12594 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17958 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51272 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7611 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14401 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
```diff
@@ -88,7 +88,7 @@ void ImageBufferBackend::convertToLuminanceMask()
 }
 
-RefPtr<PixelBuffer> ImageBufferBackend::getPixelBuffer(const PixelBufferFormat& destinationFormat, const IntRect& sourceRect, void* data, const ImageBufferAllocator& allocator) const
+RefPtr<PixelBuffer> ImageBufferBackend::getPixelBuffer(const PixelBufferFormat& destinationFormat, const IntRect& sourceRect, void* data, const ImageBufferAllocator& allocator)
```
Maybe we can keep this const if we replace `void* data` with `const void* data`.

kkinnunen-apple replied:
I don't think that is the issue nor the solution. That would make sense if all the subclasses were expected to hold the data inside their class object (the thing being mutable) inline, eg. as char m_data[700000] array.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.h:
```diff
@@ -54,7 +54,7 @@ class ImageBufferShareableMappedIOSurfaceBackend final : public WebCore::ImageBu
-    RefPtr<WebCore::NativeImage> copyNativeImage(WebCore::BackingStoreCopy = WebCore::CopyBackingStore) const final;
+    RefPtr<WebCore::NativeImage> copyNativeImage(WebCore::BackingStoreCopy = WebCore::CopyBackingStore) final;
 
     mutable WebCore::IOSurfaceSeed m_lastSeedWhenCopyingImage { 0 };
```
I think this does not need to be mutable anymore.

kkinnunen-apple committed:
ImageBufferBackend has functions marked as const when they are not that in practice
https://bugs.webkit.org/show_bug.cgi?id=253064
rdar://106020570

Reviewed by Said Abou-Hallawa.

Remove const qualifier from ImageBufferBackend methods that are not
const in practice. Even getters like context() and get*() have implicit
assumptions about lazy work and caching, so it is not useful to pretend
the methods are const.

Remove ImageBufferBackend::platformLayer() const function altogether,
it is unused.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::platformLayer const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferBackend::getPixelBuffer):
(WebCore::ImageBufferBackend::copyNativeImageForDrawing const): Deleted.
(WebCore::ImageBufferBackend::getPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::copyToPlatformTexture):
(WebCore::ImageBufferBackend::platformLayer const): Deleted.
(WebCore::ImageBufferBackend::copyToPlatformTexture const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::context):
(WebCore::ImageBufferCGBitmapBackend::backendSize const):
(WebCore::ImageBufferCGBitmapBackend::copyNativeImage):
(WebCore::ImageBufferCGBitmapBackend::getPixelBuffer):
(WebCore::ImageBufferCGBitmapBackend::context const): Deleted.
(WebCore::ImageBufferCGBitmapBackend::copyNativeImage const): Deleted.
(WebCore::ImageBufferCGBitmapBackend::getPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::context):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::context const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing const): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableBitmapBackend::getPixelBuffer):
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage const): Deleted.
(WebKit::ImageBufferShareableBitmapBackend::getPixelBuffer const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::context):
(WebKit::ImageBufferRemoteIOSurfaceBackend::copyNativeImage):
(WebKit::ImageBufferRemoteIOSurfaceBackend::getPixelBuffer):
(WebKit::ImageBufferRemoteIOSurfaceBackend::context const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::copyNativeImage const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::getPixelBuffer const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::copyNativeImage):
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::copyNativeImage const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.h:

Canonical link: https://commits.webkit.org/260988@main
